rnn_speech_recognition/pytorch/train.py

```diff
@@ -14,464 +13,562 @@
+    io.add_argument('--dali_device', type=str, choices=['cpu', 'gpu'],
+                    default='cpu', help='Use DALI pipeline for fast data processing')
+    io.add_argument('--resume', action='store_true',
+                    help='Try to resume from last saved checkpoint.')
+    io.add_argument('--ckpt', default=None, type=str,
+                    help='Path to a checkpoint for resuming training')
+    io.add_argument('--save_at_the_end', action='store_true',
+                    help='Saves model checkpoint at the end of training')
+    io.add_argument('--save_frequency', default=None, type=int,
+                    help='Checkpoint saving frequency in epochs')
+    io.add_argument('--keep_milestones', default=[], type=int, nargs='+',
+                    help='Milestone checkpoints to keep from removing')
+    io.add_argument('--save_best_from', default=200, type=int,
+                    help='Epoch on which to begin tracking best checkpoint (dev WER)')
+    io.add_argument('--val_frequency', default=1, type=int,
+                    help='Number of epochs between evaluations on dev set')
+    io.add_argument('--log_frequency', default=25, type=int,
+                    help='Number of steps between printing training stats')
+    io.add_argument('--prediction_frequency', default=None, type=int,
+                    help='Number of steps between printing sample decodings')
+    io.add_argument('--model_config', default='configs/baseline_v3-1023sp.yaml',
+                    type=str, required=True,
+                    help='Path of the model configuration file')
+    io.add_argument('--num_buckets', type=int, default=6,
+                    help='If provided, samples will be grouped by audio duration, '
+                         'to this number of backets, for each bucket, '
+                         'random samples are batched, and finally '
+                         'all batches are randomly shuffled')
+    io.add_argument('--train_manifests', type=str, required=True, nargs='+',
+                    help='Paths of the training dataset manifest file')
+    io.add_argument('--val_manifests', type=str, required=True, nargs='+',
+                    help='Paths of the evaluation datasets manifest files')
+    io.add_argument('--max_duration', type=float,
+                    help='Discard samples longer than max_duration')
+    io.add_argument('--dataset_dir', required=True, type=str,
+                    help='Root dir of dataset')
+    io.add_argument('--output_dir', type=str, required=True,
+                    help='Directory for logs and checkpoints')
+    io.add_argument('--log_file', type=str, default=None,
+                    help='Path to save the training logfile.')
+    io.add_argument('--max_symbol_per_sample', type=int, default=None,
+                    help='maximum number of symbols per sample can have during eval')
+    return parser.parse_args()
+
+
+def apply_ema(model, ema_model, decay):
```

Comment:
Its not obvious from the README that the exponential-moving-average of weights are used for the final evaluation. Might be useful to include this in the README together with the optimizer description or description of evaluation procedure. Also did the `ema` value have any influence on the convergence or eval accuracy?

Reply:
Yes, it makes the model reach better accuracy in earlier epochs. I can't find data for that now, but running training with two evaluations: one on real weights and the second on averaged weights shows the difference.

I will add a description of it to the README. Thanks for noticing it is missing.

Reply:
added here: https://github.com/mlcommons/training/pull/430/commits/78e7bf7d0f79da85d78a01ed03cdf0d51527f500#diff-a8e1aa48ee07d75e0d5e864d826a77d43be053c42641cba1be96d15648ececb8R181-R183